finality-aleph/src/sync_oracle.rs

```diff
@@ -0,0 +1,50 @@
+use std::{
+    sync::{
+        atomic::{AtomicU32, Ordering},
+        Arc,
+    },
+    time::{Duration, Instant},
+};
+
+use parking_lot::Mutex;
+use sp_consensus::SyncOracle as SyncOracleT;
+
+/// A sync oracle implementation tracking how far behind the highest known justification the node is.
+/// It defines being in major sync as knowing of any justification of an unknown block.
```

Comment:
I think this is 
1. too strong: we will be too often in the  major sync because of one unknown justification from the next block ( I envision this leading to a lot of switches between major and normal mode)
2. too weak? what about the situation when we have all of the justifications but are behind in not yet finalized branch of blocks?

Reply:
1. I'm not aware of other details but considering just block production it seems good to me that we stop producing blocks (by entering major sync) whenever there's a justified block above our top finalized, bc they'd be rejected anyway.
2. If I understand correctly this should trigger major sync, right?

Reply:
In this case the major sync state only impacts block production. Thus, my reasoning was that:
1. If we build a block on top of a block that we know is already not the best finalized block we are surely wasting work.
2. If we don't know which of the blocks will get finalized, then building on anything has a chance of doing something productive.
I think the only issue might happen during an actual catchup sync if for whatever reason we don't currently know of any future justifications. In this situation with the current definition we might end up creating some ancient useless block. I don't think this is a major problem (I might be missing something), but it would definitely require a lot of work to improve.

Reply:
1. But we dont know yet if it is not the best finalized block, we don't compare hashes and paths here, right? We can have the best block that is not finalized yet and is built on another not finalized block etc.

Reply:
And I feel like this being too strong can delay or even stop us from producing blocks.

Reply:
You mean a situation where we have the blocks, but are missing a justification in the path to finalize them? That should be really rare and fix itself very quickly. Although, if you have a decent idea that would fix that problem that would be nice.

Oh, to be clear, this is the only situation in which this can happen, since otherwise we finalize as soon as we receive the justification. There is no point at which we have both the justification and the block but we don't finalize it (other than finalization holes, as explained above).

How can this stop us from producing blocks if we are not lagging behind the finalized block (in which case our production would be wasted anyway)?

Reply:
Hm, welp, this actually doesn't work on a fresh chain, annoying... I'll try to figure out why.

Reply:
@timorleph The issue I have here is that we changed quite a bit a definition of the major sync. Before it meant that we are in the process of downloading at least `MAJOR_SYNC_BLOCKS`  (whatever that const is). Now it means we have **one** justification that we know of that is missing from us.

Reply:
Major sync state is only examined when deciding whether to produce blocks, so changing the definition only impacts that. Thus, changing the definition is definitely intended, since the new one works much better for our usecase.

Reply:
Its also used in ready endpoint (not yet used) and one place in finality-aleph (catch up fn). Im not sure if not producing a block is always better. But feel free to resolve

Reply:
Hmmmm, I'm not sure about the endpoint, but the catch-up function in finality uses the old notion of major syncing. This is not great, so I created a separate task for fixing that.